solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1211,17 +1211,22 @@ public ProcessedFilter getProcessedFilter(DocSet setFilter, List<Query> queries)
     }
 
     // Are all of our normal cached filters negative?
-    if (end > 0 && answer == null) {
-      answer = getLiveDocSet();
-    }
+    if (end > 0) {
```

Comment:
As I said, I think we shouldn't do this if equals to 1 since there will be at most one copy any way (sometimes optimizations lead to none).  Furthermore, consider that "answer" may be a SortedIntDocSet and thus upgrading it to a BitSet may be wasteful.

Reply:
Also, iff smallestIdx is >= 0 then "end" is effectively one smaller for the purposes of this check.

Reply:
> Also, iff smallestIdx is >= 0 then "end" is effectively one smaller for the purposes of this check.

not sure I understand this comment. I'll dig in further. The other changes I'll push in a few minutes.

Reply:
Pushed some of these optimizations in 5e753aa44b848f4e45361864c84be23a0c774edd - I checked if answer is instance of BitDocSet since thats the only case we can optimize easily.

Reply:
RE smallestIdx that I was saying, the logic here is confusing.  I think a refactor is in order to make this clearer.  I can do a push to your branch doing that, and thus adopting a small improvement here.  Without _something_, end could easily be equal to 1 and smallestIdx == 0 which is a scenario where we should do nothing because none of the pos/neg loops will do any work.